homeassistant/components/plum_lightpad/light.py

```diff
@@ -1,35 +1,65 @@
 """Support for Plum Lightpad lights."""
+import logging
+from typing import Callable, List
+
+from plumlightpad import Plum
+
 from homeassistant.components.light import (
     ATTR_BRIGHTNESS,
     ATTR_HS_COLOR,
     SUPPORT_BRIGHTNESS,
     SUPPORT_COLOR,
     LightEntity,
 )
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.entity import Entity
 import homeassistant.util.color as color_util
 
 from .const import DOMAIN
 
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: Callable[[List[Entity]], None],
+) -> None:
+    """Set up Plum Lightpad dimmer lights and glow rings."""
+
+    plum: Plum = hass.data[DOMAIN][entry.entry_id]
 
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
-    """Initialize the Plum Lightpad Light and GlowRing."""
-    if discovery_info is None:
-        return
+    def setup_entities(device) -> None:
+        entities = []
 
-    plum = hass.data[DOMAIN]
+        if "lpid" in device:
+            lightpad = plum.get_lightpad(device["lpid"])
+            entities.append(GlowRing(lightpad=lightpad))
 
-    entities = []
+        if "llid" in device:
+            logical_load = plum.get_load(device["llid"])
+            entities.append(PlumLight(load=logical_load))
 
-    if "lpid" in discovery_info:
-        lightpad = plum.get_lightpad(discovery_info["lpid"])
-        entities.append(GlowRing(lightpad=lightpad))
+        if entities:
+            async_add_entities(entities)
 
-    if "llid" in discovery_info:
-        logical_load = plum.get_load(discovery_info["llid"])
-        entities.append(PlumLight(load=logical_load))
+    async def new_load(device):
+        setup_entities(device)
 
-    if entities:
-        async_add_entities(entities)
+    async def new_lightpad(device):
+        setup_entities(device)
+
+    device_web_session = async_get_clientsession(hass, verify_ssl=False)
+    hass.async_create_task(
```

Comment:
I think the cleanup/shutdown hook should be at the end of this function.

Reply:
If feels wrong to close it here. 'plum' instance will call .close() for each lightpad it has. Those lightpads are kept as references in HA PlumLight and GlowRing classes for the lifetime of HA server.
Current code in production only calls Plum.cleanup() when HA server goes down, which I think is. correct. Objections?

Reply:
Calling `plum.cleanup()` isn't right here, but hooking up the Home assistant shutdown hook (that calls `plum.cleanup()` when home assistant is shutting down is.  Check your own `cleanup` function

Reply:
I am already installing the hook in __init__.py that will call plum.cleanup() on HA shutdown for the created instance of plum.
In this file the instance is not created, it is the same instance that is already registered for cleanup in __init__.py and passed here via hass.data:

plum: Plum = hass.data[DOMAIN][entry.entry_id]